bluephos/tasks/dft.py

```diff
@@ -0,0 +1,99 @@
+import os
+import tempfile
+import multiprocessing
+import pandas as pd
+import bluephos.modules.log_config as log_config
+from bluephos.modules.dft_calculators import OrcaCalculator, ASECalculator, remove_second_row
+from dplutils.pipeline import PipelineTask
+
+# Setup logging and get a logger instance
+logger = log_config.setup_logging(__name__)
+
+# DEBUG = True retains DFT details for review; False only keeps final results
+DEBUG = True
+
+# Constants
+MAX_DEFAULT_CPUS = 48  # Maximum default CPUs to use if not specified by environment
+
+
+def get_dft_calculator(package, n_cpus):
+    """
+    Get the appropriate DFT calculator based on the specified package.
+
+    Args:
+        package (str): The name of the DFT package to use ('orca' or 'ase').
+        n_cpus (int): Number of CPUs to use for the calculations.
+
+    Returns:
+        An instance of the specified DFT calculator.
+
+    Raises:
+        ValueError: If an unsupported DFT package is specified.
+    """
+
+    if package == "orca":
+        orca_path = os.path.join(os.getenv("EBROOTORCA"), "orca")
+        return OrcaCalculator(n_cpus, orca_path)
+    elif package == "ase":
+        return ASECalculator(n_cpus)
+    else:
+        raise ValueError("Unsupported DFT package")
+
+
+# Process each row of the DataFrame to perform DFT calculations
+def process_dataframe(row, t_ste, dft_calculator):
+    mol_id = row["ligand_identifier"]
+    ste = row["ste"]
+    energy_diff = row["energy diff"]
+
+    if ste is None or abs(ste) >= t_ste or energy_diff is not None:
+        logger.info(f"Skipping DFT on molecule {mol_id} based on z or t_ste conditions.")
+        return row
```

Comment:
Yes, we can create an issue to re-evaluate this.
I have some thought about it: logically, it would be ideal and elegant to allow only ligands that pass the filter to enter the next task(and we can do it). 
In practice, for the DFT task, the ideal number of nodes is around 20. Therefore, we need to allocate the full 40 CPUs if the number of valid ligands from the previous XTB step is greater than 2. Even if many invalid ligands "leak" into the DFT step, the real impact on resource utilization will be negligible. So if the number of "valid" ligands requires more CPUs than the total available on the server, the impact can be neglected.